### PR TITLE
Implement the new interface for objective functions and derivatives

### PR DESCRIPTION
**Pull request by janosg**
The changes for the objective function are described [here](https://estimagic.org/en/latest/development/eep-02-typing.html#proposal)

The changes for derivatives are described [here](https://estimagic.org/en/latest/development/eep-02-typing.html#id13)

The jax integration is postponed to a later PR as it does not entail any breaking changes or deprecations.

### CODE REVIEW

File: tests/optimagic/optimization/test_error_penalty.py [hunk 1]
```diff
@@ -54,9 +68,11 @@ def test_penalty_derivatives(func, deriv):
     partialed = functools.partial(
         func, constant=constant, slope=slope, x0=x0, dim_out=dim_out
     )
-    expected = first_derivative(partialed, x)
+    expected = first_derivative(
+        partialed, x, unpacker=lambda x: x.internal_value(solver_type)
```
Comment (timmens) [inline tests/optimagic/optimization/test_error_penalty.py:72]: Maybe we should also allow `unpacker` to be an `AggregationLevel`, in which case we set the unpacker to `lambda x: x.internal_value(unpacker)`?

File: tests/optimagic/shared/test_process_user_functions.py
Comment (timmens) [inline tests/optimagic/shared/test_process_user_functions.py:1]: level is spelled incorrectly in the entire file as levle

File: tests/optimagic/test_mark.py
Comment (timmens) [inline tests/optimagic/test_mark.py:1]: Can you add a test where a TypeError would occur (as is checked in the decorators)?

**Review by timmens — COMMENTED** [review]
Very nice PR, this is a big leap towards the new release 🎉 

I only have a few clarifying questions. Once these are discussed, I will approve.

File: docs/source/how_to/how_to_criterion_function.ipynb [hunk 2]
```diff
@@ -0,0 +1,226 @@
+{
```
Comment (timmens) [inline docs/source/how_to/how_to_criterion_function.ipynb:1]: <p>that would also work <strong>with</strong> scipy.optimize</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenSourceEconomics/optimagic/pull/512/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (timmens) [inline docs/source/how_to/how_to_criterion_function.ipynb:1]: <p>you need to <s>declare</s> mark your function with </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenSourceEconomics/optimagic/pull/512/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

File: docs/source/how_to/how_to_derivatives.ipynb [hunk 3]
```diff
@@ -0,0 +1,276 @@
+{
```
Comment (timmens) [inline docs/source/how_to/how_to_derivatives.ipynb:1]: <p><s>such</s> a derivative based</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenSourceEconomics/optimagic/pull/512/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (timmens) [inline docs/source/how_to/how_to_derivatives.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #6.        fun=sphere,
</pre><p>Do we need to pass <code>fun</code> if we pass <code>fun_and_jac</code>? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenSourceEconomics/optimagic/pull/512/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (timmens) [inline docs/source/how_to/how_to_derivatives.ipynb:1]: <p>Maybe I'd define <code>dict_params</code> at the very top. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenSourceEconomics/optimagic/pull/512/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

File: src/optimagic/deprecations.py [hunk 4]
```diff
@@ -207,6 +216,96 @@ def replace_and_warn_about_deprecated_bounds(
...
+    times and the warning would pop up too often.
+
+    """
+
+    @wraps(func)
+    def wrapper(*args: P.args, **kwargs: P.kwargs) -> Any:
+        raw = func(*args, **kwargs)
+        # fun and jac case
+        if isinstance(raw, tuple):
+            out = (convert_dict_to_function_value(raw[0]), raw[1])
```
[the remark this answers is not in the document]
Comment (janosg) [reply src/optimagic/deprecations.py:289]: Derivatives are still processed using the converter. It this moment we would not have enough information to process the derivative.

File: src/optimagic/mark.py
[the remark this answers is not in the document]
Comment (janosg) [reply src/optimagic/mark.py:1]: For most functions just adding a field works and it preserves more information than the wrapper approach. For immutable callables (like partial objects) we need the wrapper approach.

File: src/optimagic/shared/process_user_function.py [hunk 5]
```diff
@@ -27,7 +33,9 @@ def process_func_of_params(func, kwargs, name="your function", skip_checks=False
 
         raise InvalidKwargsError(msg)
 
-    out = partial(func, **kept)
+    # wraps preserves static fields that might have been added to the function via
+    # mark decorators.
+    out = wraps(func)(partial(func, **kept))
```
[the remark this answers is not in the document]
Comment (janosg) [reply src/optimagic/shared/process_user_function.py:38]: Thanks, didn't know this.

File: src/optimagic/shared/process_user_function.py [hunk 6]
```diff
@@ -1,13 +1,19 @@
...
+    LeastSquaresFunctionValue,
+    LikelihoodFunctionValue,
+    ScalarFunctionValue,
+)
+from optimagic.typing import AggregationLevel
 from optimagic.utilities import propose_alternatives
 
 
-def process_func_of_params(func, kwargs, name="your function", skip_checks=False):
+def partial_func_of_params(func, kwargs, name="your function", skip_checks=False):
```
[the remark this answers is not in the document]
Comment (janosg) [reply src/optimagic/shared/process_user_function.py:16]: Exactly, I'm not happy with the name but didn't find a better one either.

File: src/optimagic/optimization/error_penalty.py [hunk 7]
```diff
@@ -3,21 +3,29 @@
...
 
 
 def _penalty_contributions(x, constant, slope, x0, dim_out):
     contrib = (constant + slope * np.linalg.norm(x - x0)) / dim_out
-    return np.ones(dim_out) * contrib
+    contrib = np.ones(dim_out) * contrib
+    return LikelihoodFunctionValue(value=contrib)
 
 
 def _penalty_root_contributions(x, constant, slope, x0, dim_out):
```
[the remark this answers is not in the document]
Comment (janosg) [reply src/optimagic/optimization/error_penalty.py:25]: Done.

File: src/optimagic/optimization/error_penalty.py [hunk 8]
```diff
@@ -31,7 +39,7 @@ def _penalty_contributions_derivative(x, constant, slope, x0, dim_out):  # noqa:
 
 def _penalty_root_contributions_derivative(x, constant, slope, x0, dim_out):
```
[the remark this answers is not in the document]
Comment (janosg) [reply src/optimagic/optimization/error_penalty.py:40]: Done.

File: src/optimagic/optimization/internal_criterion_template.py
[the remark this answers is not in the document]
Comment (janosg) [reply src/optimagic/optimization/internal_criterion_template.py:1]: Yes, this module will need a complete rewrite, just renaming a few things would not be enough. I also didn't type it so we see that this is unfinished.

File: src/optimagic/optimization/multistart.py
[the remark this answers is not in the document]
Comment (janosg) [reply src/optimagic/optimization/multistart.py:1]: We are not simply evaluating a decorated version of the user provided `fun` but `internal_criterion_and_derivative` with the task `criterion`. So we don't get a `SpecificFunctionValue` object here. But I could get rid of the manual aggregation in a different way, by pretending we are using `internal_criterion_and_derivative` for a scalar optimizer.

[the remark this answers is not in the document]
Comment (janosg) [reply src/optimagic/optimization/multistart.py:1]: Still couldn't delete `_aggregate_func_output_to_value` but should be possible when we have the new internal algorithm interface (which will have better result objects).

File: tests/estimagic/test_estimate_ml.py [hunk 9]
```diff
@@ -116,27 +118,18 @@ def nloglikeobs(self, params, *args, **kwargs):
...
-    return loglike, loglike_derivative["value"]
-
-
 test_cases = list(
     itertools.product(
         [
-            {"algorithm": "scipy_lbfgsb"},
-            "scipy_lbfgsb",
+            # {"algorithm": "scipy_lbfgsb"},
+            # "scipy_lbfgsb",
```
[the remark this answers is not in the document]
Comment (janosg) [reply tests/estimagic/test_estimate_ml.py:125]: No, should be commented in; Did that now.

File: tests/optimagic/optimization/test_error_penalty.py (hunk 1 above)
Comment (janosg) [reply tests/optimagic/optimization/test_error_penalty.py:72]: This is mostly used internally (for example not documented in any how-to guide or tutorial) and I am hesitant to add more logic to the derivatives before the re-write.

File: tests/optimagic/shared/test_process_user_functions.py
Comment (janosg) [reply tests/optimagic/shared/test_process_user_functions.py:1]: Thanks, fixed.

File: docs/source/how_to/how_to_derivatives.ipynb (hunk 3 above)
Comment (janosg) [reply docs/source/how_to/how_to_derivatives.ipynb:1]: Currently yes but I created an issue to remove this limitation.

File: src/optimagic/shared/process_user_function.py [hunk 10]
```diff
@@ -35,7 +35,7 @@ def partial_func_of_params(func, kwargs, name="your function", skip_checks=False
 
     # wraps preserves static fields that might have been added to the function via
     # mark decorators.
```
Comment (timmens) [inline src/optimagic/shared/process_user_function.py]: Should be `update_wrapper` now.

**Review by timmens — APPROVED** [review]
Looks good, only one comment about a comment.

File: src/optimagic/shared/process_user_function.py (hunk 10 above)
Comment (timmens) [reply src/optimagic/shared/process_user_function.py]: Will be fixed in #517

**Pull request merged**